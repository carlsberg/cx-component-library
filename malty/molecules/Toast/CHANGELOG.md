# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.2.0 (2024-05-31)

### Bug Fixes

- **DSM-287:** fixed layout ([#245](https://github.com/CarlsbergGBS/cx-component-library/issues/245)) ([662a964](https://github.com/CarlsbergGBS/cx-component-library/commit/662a964748e769df36ddd0e552f5baea2522364b))
- **DSM-510:** fixed toast height ([#509](https://github.com/CarlsbergGBS/cx-component-library/issues/509)) ([d526ce7](https://github.com/CarlsbergGBS/cx-component-library/commit/d526ce7916cefc935fa0399c2e2881083d1c4453))
- Run auto fixable lint based on new `eslint` configuration ([#501](https://github.com/CarlsbergGBS/cx-component-library/issues/501)) ([9ee94cc](https://github.com/CarlsbergGBS/cx-component-library/commit/9ee94cc2cd407f717e62d8857b5f20a74e7bbec4))
- Toast bug fix ([#236](https://github.com/CarlsbergGBS/cx-component-library/issues/236)) ([e39651d](https://github.com/CarlsbergGBS/cx-component-library/commit/e39651d80024692d39038beb6e27bcf0943420f6))

### Features

- add tsconfig files and update package.json files for each component ([3bfc644](https://github.com/CarlsbergGBS/cx-component-library/commit/3bfc644e1cfc9dbb7cf7a0469e25fce055b53240))
- AlertInline and Toast ([#182](https://github.com/CarlsbergGBS/cx-component-library/issues/182)) ([8f344e3](https://github.com/CarlsbergGBS/cx-component-library/commit/8f344e392a6b339d3f159c3e8a2958ef5ddabd71))
- change organization name ([a0aa428](https://github.com/CarlsbergGBS/cx-component-library/commit/a0aa428f43138af5707a3ff4c0e36b7c056f02e6))
- **chore:** [DSM-714] add lint and format commands ([#705](https://github.com/CarlsbergGBS/cx-component-library/issues/705)) ([5b8c3bf](https://github.com/CarlsbergGBS/cx-component-library/commit/5b8c3bf45dc5c32cd52724a172578b36d93427a5))
- **DSM-22:** removed default autoTimer ([#193](https://github.com/CarlsbergGBS/cx-component-library/issues/193)) ([3a2c6ef](https://github.com/CarlsbergGBS/cx-component-library/commit/3a2c6ef53fc59e634fb90d2ef68c359113f783e5))
- **DSM-22:** Toast ([#176](https://github.com/CarlsbergGBS/cx-component-library/issues/176)) ([94faba6](https://github.com/CarlsbergGBS/cx-component-library/commit/94faba6eb4324c4975ae16dac0ddb1611cc18ea7))
- **DSM-248:** Headline HTML tags ([#184](https://github.com/CarlsbergGBS/cx-component-library/issues/184)) ([305f5e0](https://github.com/CarlsbergGBS/cx-component-library/commit/305f5e0fa92d5813f4877b338ad827a8c347d452))
- **DSM-362:** snapshots removed & unit test updates ([#307](https://github.com/CarlsbergGBS/cx-component-library/issues/307)) ([78423eb](https://github.com/CarlsbergGBS/cx-component-library/commit/78423eb5cfe04fe8a3a7af1f5af074ac6d35e235))
- **general:** [DSM-953] refactor icons implementation ([#892](https://github.com/CarlsbergGBS/cx-component-library/issues/892)) ([08d259f](https://github.com/CarlsbergGBS/cx-component-library/commit/08d259f476ce97e9db8ac41d5eb4773eabca0d37))
- **global:** [DSM-784] add space props in all components ([#842](https://github.com/CarlsbergGBS/cx-component-library/issues/842)) ([9f03f59](https://github.com/CarlsbergGBS/cx-component-library/commit/9f03f59e7518a99abcea06bacbee989d1ddb0465))
- **global:** migrate from bit to lerna ([a0110c8](https://github.com/CarlsbergGBS/cx-component-library/commit/a0110c8831370dc762c193b17cc593eed381f990))
- **global:** update storybook files to be compliant with v7 ([#970](https://github.com/CarlsbergGBS/cx-component-library/issues/970)) ([c61d4b4](https://github.com/CarlsbergGBS/cx-component-library/commit/c61d4b40a0755becf942ad3f28758a159f8c54e4))
- integrate visual testing ([#925](https://github.com/CarlsbergGBS/cx-component-library/issues/925)) ([35dfc1c](https://github.com/CarlsbergGBS/cx-component-library/commit/35dfc1cce6b76d5d2661d552f9fb25477d3a57f7))
- New Chip component and minor changes ([#239](https://github.com/CarlsbergGBS/cx-component-library/issues/239)) ([ce92ab0](https://github.com/CarlsbergGBS/cx-component-library/commit/ce92ab0543b96296952a29e2d43d0df29d4e188f))
- remove empty composition files ([f8173f4](https://github.com/CarlsbergGBS/cx-component-library/commit/f8173f4a2ecbf80bb7b6ffe848c023ae31819c2d))
- Several component updates ([#185](https://github.com/CarlsbergGBS/cx-component-library/issues/185)) ([633c9f0](https://github.com/CarlsbergGBS/cx-component-library/commit/633c9f0a4432e9e3d08ff05156453660929c19a2))
- Use new `eslint-config-carlsberg` with streamlined way to create components ([#559](https://github.com/CarlsbergGBS/cx-component-library/issues/559)) ([a459332](https://github.com/CarlsbergGBS/cx-component-library/commit/a45933215bc0b523220743f17d988d7f1ad5fce6))
- v1.0.29 minor updates ([#247](https://github.com/CarlsbergGBS/cx-component-library/issues/247)) ([73a6675](https://github.com/CarlsbergGBS/cx-component-library/commit/73a6675abfa3ff73cb0bb334d6e2cd63e6198e88))
- Version bumps for various fixes ([#205](https://github.com/CarlsbergGBS/cx-component-library/issues/205)) ([57c2e30](https://github.com/CarlsbergGBS/cx-component-library/commit/57c2e30804ba0aa6ebdd6a3ac59044b69ad9393f))
