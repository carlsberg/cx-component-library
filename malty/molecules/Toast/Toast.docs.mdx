---
labels: ['react', 'typescript', 'toast']
description: 'A Toast component.'
---
