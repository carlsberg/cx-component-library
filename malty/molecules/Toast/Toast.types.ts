export interface ToastProps {
  message: string;
  color: ToastColor;
  showCloseIcon?: boolean;
  onClose?: (state: boolean) => void;
  onCustomAction?: () => void;
  customActionText?: string;
  autoHideDuration?: number;
  dataQaId?: string;
}

export enum ToastColor {
  Notification = 'notification-strong',
  Fail = 'fail'
}
