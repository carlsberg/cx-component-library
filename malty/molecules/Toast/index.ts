export { Toast } from './Toast';
export { AlertColor, AlertSize, AlertType } from './Toast.types';
export type { AlertProps } from './Toast.types';
