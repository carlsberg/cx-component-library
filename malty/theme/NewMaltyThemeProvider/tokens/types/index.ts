export type { PrimitiveBorderRadiusTokens, SemanticBorderRadiusTokens } from './borderRadius.types';
export type { PrimitiveBorderTokens, SemanticBorderTokens } from './borders.types';
export type { PrimitiveColorsTokens, SemanticColorsTokens } from './colors.types';
export type { GridTokens } from './grid.types';
export type { OpacityTokens } from './opacity.types';
export type { ShadowTokens } from './shadows.types';
export type {
  AdditionalPrimitives,
  BorderRadiusTokens,
  ColorTokens,
  SemanticOverrides,
  SemanticTokens
} from './shared.types';
export type { SizeTokens } from './sizes.types';
export type { TypographyTokens } from './typography.types';
