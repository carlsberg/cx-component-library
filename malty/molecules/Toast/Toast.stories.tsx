import { action } from '@storybook/addon-actions';
import { Meta, Story } from '@storybook/react';
import React from 'react';
import styled from 'styled-components';
import { Toast as ToastComponent } from './Toast';
import { ToastColor, ToastProps } from './Toast.types';

export default {
  title: 'Information/Toast',
  component: ToastComponent,
  parameters: {
    importObject: 'Toast',
    importPath: '@carlsberggroup/malty.molecules.toast',
    variants: ['custom'],
  },
  argTypes: {
    message: {
      control: 'text',
      description: 'Message to be displayed',
    },
    color: {
      description: 'Toast colors, from design predefined colors, as follows.',
      options: Object.keys(ToastColor),
      mapping: ToastColor,
      control: {
        type: 'select',
        label: Object.values(ToastColor),
      },
      table: {
        defaultValue: {
          summary: 'ToastColor.Notification',
        },
      },
    },
    showCloseIcon: {
      description: 'If true close icon is displayed',
      control: 'boolean',
    },
    onClose: {
      description: 'function to be called when close icon is clicked',
    },
    customActionText: {
      control: 'text',
      description: 'Text to be diplayed as custom action',
      table: { defaultValue: { summary: 'none' } },
    },
    onCustomAction: {
      description: 'function to be called when custom action button is clicked',
    },
    autoHideDuration: {
      control: 'number',
      description: 'Set auto hide duration',
    },
    dataQaId: {
      control: 'text',
      description: 'Alert data-qi-id, can be',
      table: { defaultValue: { summary: 'none' } },
    },
  },
} as Meta;
const StyledContainer = styled.div`
  height: 200px;
  width: 100%;
  position: relative;
`;
const Template: Story<ToastProps> = ({ ...args }) => (
  <StyledContainer>
    <ToastComponent {...args} />
  </StyledContainer>
);

export const Toast = Template.bind({});

const params = new URLSearchParams(window.location.search);
const variant = params.get('variant');
switch (variant) {
  case 'custom':
    Toast.args = {
      message: 'Hello, Im the Toast Alert! Play with me.',
      color: ToastColor.Notification,
      customActionText: 'Action',
      onCustomAction: action('Custom Action clicked'),
      dataQaId: 'toast-alert',
    };
    break;
  default:
    Toast.args = {
      message: 'Hello, Im the Toast Alert! Play with me.',
      color: ToastColor.Notification,
      showCloseIcon: true,
      dataQaId: 'toast-alert',
    };
    break;
}
