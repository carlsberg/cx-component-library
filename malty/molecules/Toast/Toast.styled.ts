import styled, { keyframes } from 'styled-components';
import { ToastColor } from './Toast.types';

// Toast
const fadeInBottomUpAnimation = keyframes`
  0% {
    bottom: 0;
    opacity: 0.3
  }
  100% {
    bottom: 32px;
    opacity: 1
  }
`;
export const StyledAlertToastWrapper = styled.div`
  width: 100%;
  position: absolute;
  margin: auto;
  left: 0;
  bottom: ${({ theme }) => theme.sizes.l.value};
  animation-name: ${fadeInBottomUpAnimation};
  animation-duration: 0.7s;
  animation-timing-function: ease-in-out;
`;

export const StyledToast = styled.div<{
  color: ToastColor;
}>`
  padding: ${({ theme }) => theme.sizes.xs.value} ${({ theme }) => theme.sizes.s.value};
  display: flex;
  justify-content: space-between;
  align-items: center;
  height: ${({ theme }) => theme.sizes.xl.value};
  width: 100%;
  max-width: 736px;
  background-color: ${({ color, theme }) => theme.colors.colours.system[color].value};
  color: ${({ theme }) => theme.colors.colours.default.white.value};
  margin: auto;
  word-break: break-word;
  box-sizing: border-box;
  gap: ${({ theme }) => theme.sizes.s.value};
`;

export const StyledActionItem = styled.div`
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  box-sizing: border-box;
  button {
    text-decoration: none;
  }
`;

export const StyledDismissContainer = styled.div`
  display: flex;
  align-items: center;
  column-gap: ${({ theme }) => theme.sizes.s.value};
  margin-left: auto;
`;
