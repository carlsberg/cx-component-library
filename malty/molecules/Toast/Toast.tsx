import { Button, ButtonSize, ButtonStyle, ButtonType } from '@carlsberggroup/malty.atoms.button';
import { Text, TextColor, TextStyle } from '@carlsberggroup/malty.atoms.text';
import { Close } from '@carlsberggroup/malty.icons.close';
import { globalTheme as defaultTheme } from '@carlsberggroup/malty.theme.malty-theme-provider';
import React, { useContext, useEffect, useState } from 'react';
import { ThemeContext } from 'styled-components';
import { StyledActionItem, StyledAlertToastWrapper, StyledDismissContainer, StyledToast } from './Toast.styled';
import { ToastColor, ToastProps } from './Toast.types';

export const Toast = ({
  showCloseIcon = true,
  onClose,
  color = ToastColor.Notification,
  dataQaId,
  onCustomAction,
  customActionText,
  message,
  autoHideDuration,
  ...props
}: ToastProps) => {
  const theme = useContext(ThemeContext) || defaultTheme;

  // Toast auto hide setup
  const [isToastVisible, setToastVisible] = useState(true);
  let autoHideTimer: ReturnType<typeof setTimeout> | null = null;

  const setAutoHideTimer = () => {
    if (autoHideTimer != null) {
      clearTimeout(autoHideTimer);
    }
    if (autoHideDuration) {
      autoHideTimer = setTimeout(() => {
        hideToast();
      }, autoHideDuration) as unknown as ReturnType<typeof setTimeout>;
    }
  };

  // set timeout on component mount
  useEffect(() => {
    setAutoHideTimer();
    return () => {
      if (autoHideTimer) {
        hideToast();
        clearTimeout(autoHideTimer);
      }
    };
  }, []);

  const hideToast = () => {
    setToastVisible(false);
    if (autoHideTimer) clearTimeout(autoHideTimer);
  };

  // Action handlers
  const handleDismiss = () => {
    if (onClose) {
      onClose();
    }
    hideToast();
  };

  const handleCustomAction = () => {
    if (onCustomAction) {
      onCustomAction();
    }
    hideToast();
  };

  const renderDismissContainer = () => (
    <StyledDismissContainer theme={theme}>
      <Button
        negative
        icon={<Close />}
        size={ButtonSize.Medium}
        type={ButtonType.Default}
        style={ButtonStyle.Transparent}
        onClick={handleDismiss}
        dataTestId={`${dataQaId}-close-icon`}
      />
    </StyledDismissContainer>
  );

  const renderCustomAction = () => (
    <StyledActionItem>
      <Button
        negative
        size={ButtonSize.Medium}
        type={ButtonType.Default}
        style={ButtonStyle.Transparent}
        onClick={handleCustomAction}
        dataTestId={`${dataQaId}-custom-action`}
      >
        {customActionText}
      </Button>
    </StyledActionItem>
  );

  if (!isToastVisible) {
    return null;
  }

  return (
    <StyledAlertToastWrapper theme={theme} {...props}>
      <StyledToast data-testid={`${dataQaId}-container`} color={color} theme={theme}>
        <Text dataTestId={`${dataQaId}-message`} color={TextColor.White} textStyle={TextStyle.MediumSmallDefault}>
          {message}
        </Text>

        {customActionText && renderCustomAction()}
        {!customActionText && showCloseIcon && renderDismissContainer()}
      </StyledToast>
    </StyledAlertToastWrapper>
  );
};
