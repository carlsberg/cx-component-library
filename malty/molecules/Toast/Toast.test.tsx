import { render } from '@carlsberggroup/malty.utils.test';
import { fireEvent, screen } from '@testing-library/react';
import React from 'react';
import { Toast } from './Toast';
import { ToastColor } from './Toast.types';

describe('Toast', () => {
  it('should render Toast component with message', () => {
    const dataQaId = 'toast';
    render(<Toast message="Toast message" color={ToastColor.Notification} dataQaId={dataQaId} />);

    expect(screen.getByText('Toast message')).toBeInTheDocument();
    expect(screen.getByTestId(`${dataQaId}-message`)).toBeInTheDocument();
  });

  it('Should call On close action', async () => {
    const dismissMock = jest.fn();
    const dataQaId = 'toast';
    const message = 'Toast with close icon';

    render(
      <Toast
        message={message}
        showCloseIcon
        color={ToastColor.Notification}
        dataQaId={dataQaId}
        onClose={dismissMock}
      />
    );

    expect(screen.getByText(message)).toBeInTheDocument();

    fireEvent.click(screen.getByTestId(`${dataQaId}-close-icon`));

    expect(dismissMock).toHaveBeenCalledTimes(1);
  });

  it('Should call custom action', async () => {
    const dismissMock = jest.fn();
    render(
      <Toast
        message="Toast message"
        color={ToastColor.Notification}
        onCustomAction={dismissMock}
        customActionText="action"
      />
    );

    expect(screen.getByText('action')).toBeInTheDocument();
  });
});
