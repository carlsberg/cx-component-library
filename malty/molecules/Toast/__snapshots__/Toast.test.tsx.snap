// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast matches snapshot 1`] = `
<div
  className="sc-kDDrLX jwOBGr"
>
  <div
    className="sc-iqcoie eyjmYd"
    color="notification-strong"
    data-testid="toast-container"
  >
    <span
      className="sc-bczRLJ hROJNc"
    >
      <p
        className="sc-dkzDqf jdtZuS"
        color="white"
        data-testid="toast-message"
      >
        Toast message
      </p>
    </span>
    <div
      className="sc-evZas cWbSUu"
    >
      <span
        className="sc-bczRLJ hROJNc"
      >
        <button
          className="sc-gKXOVf sc-papXJ hQuGSn fESpCB"
          data-testid="toast-close-icon"
          fontSize="16px"
          onClick={[Function]}
          tabIndex={-1}
          type="button"
        >
          <div
            className="text-container "
          >
            <svg
              className="sc-gsnTZi gIwTyE"
              color="#212833"
              data-testid="svg-component"
              size="24px"
              version="1.1"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
              xmlnsXlink="http://www.w3.org/1999/xlink"
            >
              <g
                fillRule="evenodd"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M6.249 4.846l.094.083L12 10.585l5.657-5.656a1 1 0 011.497 1.32l-.083.094L13.415 12l5.656 5.657a1 1 0 01-1.32 1.497l-.094-.083L12 13.415l-5.657 5.656a1 1 0 01-1.497-1.32l.083-.094L10.585 12 4.929 6.343a1 1 0 011.32-1.497z"
                />
              </g>
            </svg>
          </div>
        </button>
      </span>
    </div>
  </div>
</div>
`;
